utils/src/Vulkan/Utils/Reflect.hs

```diff
@@ -0,0 +1,567 @@
+makePipelineShaderStageCreateInfos stage entryPoints shaderModule =
+  V.map (makePipelineShaderStageCreateInfo stage shaderModule) entryPoints
+
+-- https://github.com/KhronosGroup/GLSL/blob/master/extensions/khr/GL_KHR_vulkan_glsl.txt
+-- https://vulkan.lunarg.com/doc/view/1.2.135.0/linux/tutorial/html/08-init_pipeline_layout.html
+makeUboDescriptorSetLayoutBinding
+  :: ShaderStage -> Ubo -> (Int, DescriptorSetLayoutBinding)
+makeUboDescriptorSetLayoutBinding stage Ubo {..} =
+  ( set
+  , zero { binding         = fromIntegral binding
+         , descriptorType  = DESCRIPTOR_TYPE_UNIFORM_BUFFER
+         , descriptorCount = maybe 1 (fromIntegral . V.sum) array
+         , stageFlags      = shaderStageFlagBits stage
+         }
+  )
+
+data TextureDescriptorType = Sampler2D
+  deriving Show
+
+textureDescriptorTypeFromText :: Text -> Maybe TextureDescriptorType
+textureDescriptorTypeFromText = \case
+  "sampler2D" -> Just Sampler2D
+  _           -> Nothing
+
+textureDescriptorTypeToText :: TextureDescriptorType -> Text
+textureDescriptorTypeToText = \case
+  Sampler2D -> "sampler2D"
+
+instance FromJSON TextureDescriptorType where
+  parseJSON = withTextMaybe "type" textureDescriptorTypeFromText
+
+instance ToJSON TextureDescriptorType where
+  toJSON = String . textureDescriptorTypeToText
+
+convertTextureDescriptorType :: TextureDescriptorType -> DescriptorType
+convertTextureDescriptorType = \case
+  Sampler2D -> DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER
+
+makeTextureDescriptorSetLayoutBinding
+  :: ShaderStage -> Texture -> (Int, DescriptorSetLayoutBinding)
+makeTextureDescriptorSetLayoutBinding stage Texture {..} =
+  ( set
+  , zero { binding         = fromIntegral binding
+         , descriptorType  = convertTextureDescriptorType type'
+         , descriptorCount = maybe 1 (fromIntegral . V.sum) array
+         , stageFlags      = shaderStageFlagBits stage
+         }
+  )
+
+makeDescriptorSetLayoutBindings
+  :: ShaderStage
+  -> Vector Ubo
+  -> Vector Texture
+  -> Vector (Int, DescriptorSetLayoutBinding)
+makeDescriptorSetLayoutBindings stage ubos textures =
+  uboBindings <> textureBindings
+ where
+  uboBindings = V.map (makeUboDescriptorSetLayoutBinding stage) ubos
+  textureBindings =
+    V.map (makeTextureDescriptorSetLayoutBinding stage) textures
+
+makeDescriptorSetLayoutCreateInfos
+  :: Vector (Int, DescriptorSetLayoutBinding)
+  -> V.Vector (DescriptorSetLayoutCreateInfo '[])
+makeDescriptorSetLayoutCreateInfos bindings =
+  V.map makeDescriptorSetLayoutCreateInfo
+    . V.fromList
+    . M.elems
+    . M.unionWith (V.++) sets
+    $ setsMap
+ where
+  setSize = V.maximum . (fst <$>) $ bindings :: Int
+  sets :: Map Int (Vector DescriptorSetLayoutBinding)
+  sets = M.fromList . map (, []) $ [0 .. setSize]
+  setsMap :: Map Int (Vector DescriptorSetLayoutBinding)
+  setsMap =
+    M.fromList
+      . map (liftA2 (,) (fst . head) (V.fromList . (snd <$>)))
+      . (NE.toList <$>)
```

Comment:
It is too early to drop `NonEmpty` here, the key function can still benefit from using `NE.head`.

And I'd still prefer the map body extracted so the _bindings-to-Map_ pipeline would contain only the structural changes.
Then the _group-to-pair_ function can have its argument matched and passed without `liftA2` and pointfree.

It's a bit more verbose, yes, but makes the data flow more legible.

Reply:
thanks for reviewing, fixed by add a local function `extract :: NE.NonEmpty (Int, DescriptorSetLayoutBinding) -> (Int, Vector DescriptorSetLayoutBinding)`